src/hotspot/share/runtime/vmStructs.cpp

```diff
@@ -1981,31 +1981,31 @@ extern "C" {
 #define STRIDE(array) ((char*)&array[1] - (char*)&array[0])
 
 JNIEXPORT VMStructEntry* gHotSpotVMStructs = VMStructs::localHotSpotVMStructs;
-JNIEXPORT uint64_t gHotSpotVMStructEntryTypeNameOffset = offset_of(VMStructEntry, typeName);
-JNIEXPORT uint64_t gHotSpotVMStructEntryFieldNameOffset = offset_of(VMStructEntry, fieldName);
-JNIEXPORT uint64_t gHotSpotVMStructEntryTypeStringOffset = offset_of(VMStructEntry, typeString);
-JNIEXPORT uint64_t gHotSpotVMStructEntryIsStaticOffset = offset_of(VMStructEntry, isStatic);
-JNIEXPORT uint64_t gHotSpotVMStructEntryOffsetOffset = offset_of(VMStructEntry, offset);
-JNIEXPORT uint64_t gHotSpotVMStructEntryAddressOffset = offset_of(VMStructEntry, address);
+extern JNIEXPORT constexpr uint64_t gHotSpotVMStructEntryTypeNameOffset = offsetof(VMStructEntry, typeName);
```

Comment:
I got a partial understanding about this change, I thought having `JNIEXPORT` and `constexpr` together wouldn't work. But the symbol is still exported:
```
objdump --syms build/gcc/jdk/lib/server/libjvm.so | grep gHotSpotVMTypeEntryIsIntegerTypeOffset
000000000137ca78 g     O .rodata	0000000000000008              gHotSpotVMTypeEntryIsIntegerTypeOffset
```
and I tested that the values are still readable. I was wondering if this would be portable?

Reply:
I believe constexpr means the entity marked as a constant expression will be treated as such if conditions allow it to be used as a compile time constant, and doesn't mean it will disappear from the final binary entirely. Since JNIEXPORT exports the symbol from HotSpot, it is obviously used and the compiler takes note of that, hence allowing it to be JNIEXPORT despite being constexpr

Also I think the JNIEXPORT would preferably be in front of the extern (Actually, could I ask why extern is required here?)